src/Spatial/Internals/ImmutableList.cs

```diff
@@ -0,0 +1,76 @@
+﻿namespace MathNet.Spatial.Internals
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+
+    internal sealed class ImmutableList<T> : IEnumerable<T>
+    {
+        internal static readonly ImmutableList<T> Empty = new ImmutableList<T>();
+
+        private readonly T[] data;
+
+        private ImmutableList()
+        {
+            this.data = new T[0];
+        }
+
+        internal ImmutableList(T[] data)
+        {
+            this.data = data;
+        }
+
+        public int Count => this.data.Length;
+
+        public T this[int index] => this.data[index];
+
+        public IEnumerator<T> GetEnumerator() => ((IList<T>)this.data).GetEnumerator();
+
+        IEnumerator IEnumerable.GetEnumerator() => this.data.GetEnumerator();
+
+        internal ImmutableList<T> Add(T value)
```

Comment:
I assume most of the lists will be small and nobody will be doing `Add` and `Remove` in the loop

Reply:
We can add an AddRange method